### ISSUE TITLE
Plugin installation fails with ModuleNotFoundError on Windows

### ISSUE THREAD

**Issue by barnumbirr** [naclander/Calibre-Remarkable-Device-Driver-Plugin#1]
Hello,

just tried installing `v1.01` of the plugin to a clean install of `calibre v5.25.0` with little success:

```bash
calibre, version 5.25.0
ERROR: Unhandled exception: <b>ModuleNotFoundError</b>:No module named 'paramiko'

calibre 5.25 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 1, 0)
Traceback (most recent call last):
  File "calibre\gui2\preferences\plugins.py", line 327, in add_plugin
  File "calibre\customize\ui.py", line 477, in add_plugin
  File "calibre\customize\ui.py", line 65, in load_plugin
  File "calibre\customize\zipplugin.py", line 293, in load
  File "importlib\__init__.py", line 127, in import_module
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "calibre\customize\zipplugin.py", line 192, in exec_module
  File "calibre_plugins.remarkable_plugin.__init__", line 12, in <module>
ModuleNotFoundError: No module named 'paramiko'

```
Please let me know if I can help with anything.
Cheers.

**Comment by naclander** [comment]
Hey! Thanks for using the plugin! 

Looks like a dependency of this plugin ( that I forgot about ) is `paramiko`. Looking over the package I think you will also need `fusepy`. I initially bundled `remarkable-fs` inside this plugin because I thought it was the only dependency which I'm now realizing is incorrect.

To immediately unblock yourself, you can run `sudo pip install git+https://github.com/nick8325/remarkable-fs` which will install `remarkable-fs `( and all other dependencies like `paramiko`, and `fusepy` ). Unfortunately the master branch of `remarkable-fs` is not what is currently in `pypy`, which is why I provided the pip install from git command.

I need to figure out if I want to tell users to have all dependencies installed on their system before running the plugin, or if I want to bundle all dependencies in the zip archive. 

Can you let me know if this plugin works for you once you've ran the above pip command?

**Comment by barnumbirr** [comment]
Hey,

thanks for the quick reply. As calibre ships it's own, system independant version of Python (called `embedded-python`), it doesn't look like installing `remarkable-fs` from master helps as the same error still crops up.
Guess calibre plugins do have to ship with all bundled dependencies.

Cheers.

**Comment by naclander** [comment]
Hmm there might be a difference between Windows ( it looks like you're on Windows ) and Linux then. I've only tested and developed this plugin on Linux and while developing it, I was using my system installed version of `paramiko`, `fusepy`, and `remarkable-fs` ( which installed both of those for me ). 

It could be the case that on Windows installations the python dependencies are all required to be bundled. But this doesn't make sense to me, as how would `remarkable-fs` know to import a dependency from a local folder, instead of from the system? 

I'm new to Calibre plugins in general, so my understanding may not be complete. I'm not actually sure if bundling all plugins inside the archive will actually solve your problem in this case ( what I think will happen is that `remarkable-fs` will crash because it can't find its `paramiko` dependency, for example ). Do you know how other Calibre plugins manage this for windows? Can you verify your system python or that `embedded-python` binary can import `remarkable-fs` or `paramiko`?

**Comment by barnumbirr** [comment]
Trying to import `remarkable-fs` or `paramiko` from calibre's `embedded-python` fails. This makes sense as neither were installed: `remarkable-fs` fails to install as `paramiko` is missing and `paramiko` is well, missing. 😂 

```bash
C:\Users\[user]>calibre-debug -c "import remarkable-fs"
Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\debug.py", line 272, in main
  File "<string>", line 1
    import remarkable-fs
                     ^
SyntaxError: invalid syntax

C:\Users\[user]>calibre-debug -c "import paramiko"
Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\debug.py", line 272, in main
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'paramiko'
```
If I manually install `remarkable-fs` on my local system or through WSL, I still can't import any of the libraries as they're not installed in calibre's Python environment. Running `pip/easy_install` in said environment doesn't work either (SyntaxError).

The only other plugin I use (on both Linux and Windows) is [DeDRM](https://github.com/apprenticeharper/DeDRM_tools).

**Comment by naclander** [comment]
So I didn't quite understand what DeDRM did for its pycrypto dependency, but I did find the following things:

1) The Calibre author just says ["package all the deps youtself"](https://www.mobileread.com/forums/showthread.php?t=262052) but I still don't know how to package the dependencies of the dependencies, and make sure all the import paths work

2) [This plugin](https://github.com/xxyzz/WordDumb/blob/master/unzip.py#L86) actually does a full pip install as part of startup which I guess could work with this plugin. 

I still think this is only a problem on Windows. I'm pretty sure the Calibre plugins just use the Linux system python packages if the import is called, and I suspect the same is true for mac os.

So I guess the next step is to find a way to add all these dependencies to the archive, or have them all get installed at startup with the least amount of work.

**Comment by naclander** [comment]
Ok so I ended up trying to bundle all the dependencies in the archive, and then adding that dependency directory to the path so that imports should work. You can see the diff [here](https://github.com/naclander/Calibre-Remarkable-Device-Driver-Plugin/commit/ca2a45708320bea726cc911bf59371f35702614d)

I've created a test plugin for this and attached it to this comment. Can you let me know if this works on your machine?
[calibre-remarkable-device-driver-plugin.zip](https://github.com/naclander/Calibre-Remarkable-Device-Driver-Plugin/files/6988585/calibre-remarkable-device-driver-plugin.zip)

**Comment by barnumbirr** [comment]
This now fails to load `remarkable-fs`:

```bash
calibre, version 5.25.0
ERROR: Unhandled exception: <b>ModuleNotFoundError</b>:No module named 'remarkable_fs'

calibre 5.25 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 1, 0)
Traceback (most recent call last):
  File "calibre\gui2\preferences\plugins.py", line 327, in add_plugin
  File "calibre\customize\ui.py", line 477, in add_plugin
  File "calibre\customize\ui.py", line 65, in load_plugin
  File "calibre\customize\zipplugin.py", line 293, in load
  File "importlib\__init__.py", line 127, in import_module
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "calibre\customize\zipplugin.py", line 192, in exec_module
  File "calibre_plugins.remarkable_plugin.__init__", line 18, in <module>
ModuleNotFoundError: No module named 'remarkable_fs'
```

**Comment by naclander** [comment]
Well I guess firstly I'm surprised it fails differently in those two scenarios. It probable that I'm not understanding the Calibre plugin load process. I need to a closer look at this. Its a bit tricky since I don't have a windows machine to test out my changes.

**Comment by naclander** [comment]
I'm going to try and get hold of a Windows machine so I can replicate your error on my end. This should make debugging easier on my end.

**Comment by naclander** [comment]
So turns out this wasn't a windows specific issue. You can look at the latest set of commits if you're curious about the fix. But tl;dr all dependencies should now be included properly in the zip file. Windows is also officially supported now.

You can download the latest release here: https://github.com/naclander/Calibre-Remarkable-Device-Driver-Plugin/releases/tag/1.1.0

Let me know if this fixed the issue for you.

**Comment by barnumbirr** [comment]
Installation now works a treat, awesome job 💪
I now have to figure out how to get books onto my Remarkable as I can't seem to "Send to device" even though the plugin has been configured correctly.
Thanks for taking the time to look into the issue and for the plugin 😉

**Comment by naclander** [comment]
Please open another issue if you're having trouble connecting to the device. A couple things that might help:

Run calibre in debug mode ( not sure how to do that on windows, on Linux you need to run `calibre-debug -g` to launch calibre.
Make sure you can `ssh` into your calibre device.
Open the Plugin settings and make sure all the information is correct ( the default values are probably not right for your setup )
